extensions/amp-video/0.1/mixins.js

```diff
@@ -0,0 +1,524 @@
+
+    /** @private {boolean} */
+    this.isAllocated_ = false;
+
+    /** @private {?../../amp-story/0.1/media-pool.MediaPool} */
+    this.pool_ = null;
+
+    /** @private {?../../amp-story/0.1/media-pool.MediaInfoDef} */
+    this.mediaInfo_ = EMPTY_MEDIA_INFO;
+
+    /** @private @const {!Promise<!../../amp-story/0.1/media-pool.MediaPool>} */
+    this.poolPromise_ = Services.mediaPoolFor(element).then(service =>
+      (this.pool_ = service.poolFor(element)));
+  }
+
+  /** @override */
+  isVideoSupported() {
+    return true; // Assumption.
+  }
+
+  /** @override */
+  installEventHandlers(unusedEventsToForward, setMuted) {
+    // No need to forward events as MediaPool re-dispatches all playback events
+    // from parent element.
+
+    const {element} = this.impl;
+
+    listen(element, MediaPoolEvents.ALLOCATED, () => {
+      this.isAllocated_ = true;
+      this.hidePlaceholder();
+    });
+
+    listen(element, MediaPoolEvents.DEALLOCATED, e => {
+      const {detail} = e;
+      this.isAllocated_ = false;
+      this.mediaInfo_ =
+        /** @type {!../../amp-story/0.1/media-pool.MediaInfoDef} */ (detail);
+      this.showPlaceholder();
+    });
+
+    listen(element, 'volumechange', () => {
+      setMuted(this.getMediaInfo_('muted'));
+    });
+  }
+
+  /** @visibleForTesting */
+  showPlaceholder() {
+    if (this.placeholder_) {
+      this.togglePlaceholder_(/* show */ true);
+      return this.placeholder_;
+    }
+
+    const {element} = this.impl;
+    const posterSrc = dev().assertString(element.getAttribute('poster'));
+
+    const img = new Image();
+    const {impl} = this;
+    const {win} = this.impl;
+
+    const placeholder = createElementWithAttributes(win.document, 'div', dict({
+      'class': 'i-amphtml-placeholder',
+    }));
+
+    img.src = posterSrc;
+
+    this.placeholder_ = impl.loadPromise(img).then(() =>
+      impl.getVsync().mutatePromise(() => {
+        setImportantStyles(placeholder, {
+          'background-image': `url(${posterSrc})`,
+          'background-size': 'cover',
+          'pointer-events': 'none',
+        });
+        impl.applyFillContent(placeholder);
+        element.appendChild(placeholder);
+        return placeholder;
+      }));
+
+    return this.placeholder_;
+  }
+
+  /** @visibleForTesting */
+  hidePlaceholder() {
+    this.togglePlaceholder_(/* show */ false);
+  }
+
+  /**
+   * @param {boolean} show
+   * @private
+   */
+  togglePlaceholder_(show) {
+    if (!this.placeholder_) {
+      return;
+    }
+    const {impl} = this;
+
+    this.placeholder_.then(placeholder =>
+      impl.getVsync().mutate(() => {
+        if (show) {
+          resetStyles(placeholder, ['display']);
+          return;
+        }
+        this.placeholder_.then(placeholder =>
+          impl.getVsync().mutate(() => {
+            setImportantStyles(placeholder, {'display': 'none'});
+          }));
+      }));
+  }
+
+  /** @override */
+  whenLoadStarts() {
```

Comment:
I think it will be useful for us to be able to differentiate between "the poster has loaded" and "the video has loaded".  Is there a reasonable way to do that with this API?  (Separate PR is ok)